src/realm/object_converter.cpp

```diff
@@ -253,29 +244,210 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
+
+        // if ()
+        //     copy_list(*n_src_list, *n_dst_list, update_out);
+    };
+    auto handle_dict_copy = [&, this](StringData key) {
+        dst_dict.insert_collection(key, CollectionType::Dictionary);
+        auto n_src_dict = src_dict.get_dictionary(key);
+        auto n_dst_dict = dst_dict.get_dictionary(key);
+        copy_dictionary_in_mixed(*n_src_dict, *n_dst_dict, update_out);
+        // if () {
+        //     copy_dictionary(*n_src_dict, *n_dst_dict, update_out);
+        // }
+    };
+
+    // main algorithm
+    auto n = src_dict.size();
+    for (size_t i = 0; i < n; ++i) {
+        auto [key, any] = src_dict.get_pair(i);
+        if (any.is_type(type_List)) {
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_List);
+            handle_list_copy(key.get_string());
+        }
+        else if (any.is_type(type_Set)) {
+            // nested collections do not support nested sets, set can only be a terminal leaf
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Set);
+            handle_set_copy(key.get_string());
+        }
+        else if (any.is_type(type_Dictionary)) {
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Dictionary);
+            handle_dict_copy(key.get_string());
+        }
+        else {
+            // copy single element
+            InterRealmValueConverter::ConversionResult converted_src;
+            auto dst_obj = dst_dict.get_obj();
+            auto dst_col = dst_dict.get_col_key();
+            auto dst_mixed = dst_obj.get_any(dst_col);
+            if (cmp_src_to_dst(any, dst_mixed, &converted_src, update_out)) {
+                dst_dict.insert(key, converted_src.converted_value);
+            }
+            // return true;
+        }
+    }
+    return false;
+}
+
+void InterRealmValueConverter::handle_collection_in_mixed_mismatch(Mixed src, Mixed dst, DataType type)
+{
+    auto type_to_string = [](Mixed mixed) {
+        if (mixed.is_type(type_List))
+            return "type_List";
+        if (mixed.is_type(type_Set))
+            return "type_Set";
+        if (mixed.is_type(type_Dictionary))
+            return "type_Dictionary";
+        throw InvalidArgument("type must be List/Set/Dictionary");
+    };
+    auto is_null = dst.is_null();
+    if (!is_null && !dst.is_type(type)) {
+        // we might be trying to turn a set into a list or a dictionary. Although conversations from set to list
+        // are pretty easy, converting from or to a dictionary is not that straight foward, for this reason it is
+        // better to stop any attempt and throw an exception if we realise that src and dst collection do not match.
+        // Note: this can happen at any level in the nesting collection tree.
+        throw InvalidArgument(
```

Comment:
Why does this have to be a fatal error?
For example, in DiscardLocal mode, what would be the consequence of getting a client reset that changes the nested structure type from a list to a dictionary for example? Isn't the user/app code supposed to handle all the different schema types now?

Reply:
this should now be solved.